heron/schedulers/src/java/com/twitter/heron/scheduler/aurora/AuroraCLIController.java

```diff
@@ -112,10 +123,122 @@ public void removeContainers(Set<PackingPlan.ContainerPlan> containersToRemove)
+          "empty resource per container query output by Aurora");
+    }
+
+    String jsonStr = stderr.toString();
+    LOG.info(String.format("Resource per container: %s", jsonStr));
+    ObjectMapper mapper = new ObjectMapper();
+    JsonNode actualObj = mapper.readTree(jsonStr);
+    return actualObj;
+  }
+
+  private boolean hasEnoughQuota(JsonNode resource, Integer count)
+      throws JsonProcessingException, IOException {
+    // calc available quota
+    JsonNode quota = askAuroraQuota();
+
+    JsonNode quotaNode = quota.get("quota");
+    long quotaDiskMb = quotaNode.get("diskMb").longValue();
+    double quotaNumCpus = quotaNode.get("numCpus").doubleValue();
+    long quotaRamMb = quotaNode.get("ramMb").longValue();
+
+    JsonNode prodNode = quota.get("prodSharedConsumption");
+    long prodDiskMb = prodNode.get("diskMb").longValue();
+    double prodNumCpus = prodNode.get("numCpus").doubleValue();
+    long prodRamMb = prodNode.get("ramMb").longValue();
+
+    long availableDiskMb = quotaDiskMb - prodDiskMb;
+    double availableNumCpus = quotaNumCpus - prodNumCpus;
+    long availableRamMb = quotaRamMb - prodRamMb;
+
+    // calc res for one container
+    JsonNode containerNode =
+        resource.get(0).get("active").get(0).get("assignedTask").get("task").get("resources");
+    double containerNumCpus = containerNode.get("numCpus").doubleValue();
+    long containerRamMb = containerNode.get("ramMb").longValue();
+    long containerDiskMb = containerNode.get("diskMb").longValue();
+
+    // check if res is enough
+    if (containerNumCpus * count > availableNumCpus) {
+      String msg =
+          String.format("not enough cpu quota: %f cpu per container x %d container > %f available",
+              containerNumCpus, count, availableNumCpus);
+      throw new TopologyUpdateRecoverableException(msg);
+    } else if (containerRamMb * count > availableRamMb) {
+      String msg =
+          String.format("not enough ram quota: %d ram per container x %d container > %d available",
+              containerRamMb, count, availableRamMb);
+      throw new TopologyUpdateRecoverableException(msg);
+    } else if (containerDiskMb * count > availableDiskMb) {
+      String msg = String.format(
+          "not enough disk quota: %d disk per container x %d container > %d available",
+          containerDiskMb, count, availableDiskMb);
+      throw new TopologyUpdateRecoverableException(msg);
+    }
+    return true;
+  }
+
+  private boolean isProd(JsonNode jsonStatus) {
+    String tier = jsonStatus.get(0).get("active").get(0).get("assignedTask").get("task").get("tier")
+        .textValue();
+    if ("preferred".equals(tier)) {
+      return true;
+    }
+    return false;
+  }
+
   @Override
   public Set<Integer> addContainers(Integer count) {
-    //aurora job add <cluster>/<role>/<env>/<name>/<instance_id> <count>
-    //clone instance 0
+    try {
+      JsonNode jsonStatus = askAuroraStatus();
```

Comment:
We can simply get the env from `Context` (https://github.com/twitter/heron/blob/master/heron/spi/src/java/com/twitter/heron/spi/common/Context.java#L36). No need to ask Aurora Status.

Reply:
i do not think so. the aurora tier is not same with env in job spec key

Reply:
I mean, only when users try to update a topology in `prod` env, which can be parsed from the `Context`, we will try to `askAuroraQuota()`; we don't need the method `askAuroraStatus()`

Reply:
It is not a safe assumption.

Reply:
@nwangtw it should be safe; this should be verified ahead to come to the actual update stage.
Furthermore, topology-name is unique for a cluster, so there will be not topologies with the same name in one cluster.

Reply:
I mean from aurora point of view.

From aurora doc it seems to me that environments are configurable. For example, inside Twitter we call it "prod" and in apache doc it is called "production".

Another one is that users might config their prod jobs to be preemptible if the SLA is low, which doesn't need quota although it is a "prod" job. This requires user to customize the aurora configs, but it is quite easy from heron design.